### PR TITLE
feat: ValidateIf decorator

### PR DESCRIPTION
**Pull request by edcarroll**
This PR covers the addition of a potential @ValidateIf() decorator.

Definition is as follows:
```ts
@ValidateIf((object:any) => boolean)
```

The decorator can be attached to any property with validation rules, and is run by passing the object being validated to the constraint function, which returns a boolean specifying whether or not to validate that property.

Example:
```ts
export class Test {
    public conditionalProp:string;

    @ValidateIf((o:any) => o.conditionalProp == "value")
    @isNotEmpty()
    public example:string;
}

let t = new Test()
t.conditionalProp = "other";
let errors = validate(t); // -> produces error as 'example' is not empty.
t.conditionalProp = "value";
errors = validate(t); // -> produces no errors.
```

Let me know your thoughts on this, and if this is a feature you would like to add I will write tests and update the release notes.

### CODE REVIEW

**Comment by pleerock** [comment]
tests are failing

File: src/decorator/decorators.ts [hunk 1]
```diff
@@ -63,6 +63,22 @@ export function ValidateNested(validationOptions?: ValidationOptions) {
     };
 }
 
+/**
+ * Objects / object arrays marked with this decorator will also be validated.
+ */
+export function ValidateIf(condition: (object: any, value: any) => boolean, validationOptions?: ValidationOptions) {
```
Comment (pleerock) [inline src/decorator/decorators.ts]: not sure if I like this name `ValidateIf`. Specifically I don't like `Validate` word here. Maybe we can come up with better name?

Comment (pleerock) [inline src/decorator/decorators.ts]: no, take my words back, I just got this decorator's purpose

File: src/validation/ValidationExecutor.ts [hunk 2]
```diff
@@ -50,10 +50,16 @@ export class ValidationExecutor {
             const metadatas = groupedMetadatas[propertyName].filter(metadata => metadata.type !== ValidationTypes.IS_DEFINED);
             const customValidationMetadatas = metadatas.filter(metadata => metadata.type === ValidationTypes.CUSTOM_VALIDATION);
             const nestedValidationMetadatas = metadatas.filter(metadata => metadata.type === ValidationTypes.NESTED_VALIDATION);
+            const conditionalValidationMetadatas = metadatas.filter(metadata => metadata.type == ValidationTypes.CONDITIONAL_VALIDATION);
 
             const validationError = this.generateValidationError(object, value, propertyName);
             validationErrors.push(validationError);
 
+            let canValidate = this.conditionalValidations(object, value, conditionalValidationMetadatas);
```
Comment (pleerock) [inline src/validation/ValidationExecutor.ts]: use `const` instead

File: src/validation/ValidationExecutor.ts [hunk 3]
```diff
@@ -111,6 +117,14 @@ export class ValidationExecutor {
         return validationError;
     }
 
+    private conditionalValidations(object: Object,
+                               value: any,
+                               metadatas: ValidationMetadata[]) {
```
Comment (pleerock) [inline src/validation/ValidationExecutor.ts]: bad code formatting

**Comment by pleerock** [comment]
please update all docs in readme and provide a tests for this new decorator

**Comment by edcarroll** [comment]
I've updated the docs and added tests, let me know if you need anything else :smile:

**Comment by pleerock** [comment]
Looks good now, thank you

**Comment by pleerock** [comment]
just released a highly build, you can use `npm i class-validator@next` to install it

**Comment by edcarroll** [comment]
Brilliant, will switch to that tomorrow! Thanks

**Comment by sanex3339** [comment]
amazing!

**Comment by pleerock** [comment]
0.6.0 is released with this feature

**Comment by sanex3339** [comment]
Amazing. As soon as i have need in this feature - this feature was PRed and released.

**Pull request merged**